File: 442-find-all-duplicates-in-an-array/find-all-duplicates-in-an-array.cpp
class Solution {
public:
    vector<int> findDuplicates(vector<int>& nums) {
        int i = 0;
        int n = nums.size();

        while(i<n){
            int correctindex = nums[i]-1;
            if(nums[i]!=nums[correctindex]){
                int temp = nums[i];
                nums[i]=nums[correctindex];
                nums[correctindex]=temp;
            }
            else{i++;}
        }

        vector <int>ans;
        for(int j=0;j<n;j++){
            if(nums[j]!=j+1){
                ans.push_back(nums[j]);
            }
        }
        return ans;
    }
};

File: 442-find-all-duplicates-in-an-array/find-all-duplicates-in-an-array.py
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        i = 0
        n = len(nums)

        while i < n:
            correct_index = nums[i] - 1

            if nums[i] != nums[correct_index]:
                nums[i], nums[correct_index] = nums[correct_index], nums[i]
            else:
                i += 1

        ans = []
        for j in range(n):
            if nums[j] != j + 1:
                ans.append(nums[j])

        return ans
